File: src/187_Repeated_DNA_Sequences.rs
// 187_Repeated_DNA_Sequences
// https://leetcode.cn/problems/repeated-dna-sequences/

use std::collections::HashMap;
impl Solution {
    pub fn find_repeated_dna_sequences(s: String) -> Vec<String> {
        if s.len() < 10 {
            return vec![];
        }
        let mut map: HashMap<&str, usize> = HashMap::new();
        let mut ans = vec![];
        let window = 10 as usize;
        for i in 0..s.len()-window+1 {
            let sub = &s[i..i+window];
            map.entry(sub).and_modify(|cnt| {
                *cnt += 1;
                if *cnt == 2 {
                    ans.push(sub.to_string());
                }
            }).or_insert(1);
        }
        ans
    }
}

/// 示例代码
impl Solution {
    pub fn find_repeated_dna_sequences(s: String) -> Vec<String> {
        use std::collections::HashMap;

        let mut map = HashMap::new();
        if s.len() < 10 {
            return vec![];
        }

        let sv = s.as_str();
        for i in 0..=(s.len() - 10) {
            map.entry(&s[i..(i+10)]).and_modify(|cnt| *cnt += 1).or_insert(1);
        }
        map.iter().filter(|(_k, &v)|v > 1).map(|(k, _v)| k.to_string()).collect()
    }
}


use std::collections::HashSet;
impl Solution {
    pub fn find_repeated_dna_sequences(s: String) -> Vec<String> {
        if s.len() < 10 {
            return vec![];
        }

        let mut set: HashSet<Vec<char>> = HashSet::new();
        let mut ans1: HashSet<Vec<char>> = HashSet::new();
        let s: Vec<char> = s.chars().collect();
        let mut left = 0;
        let mut right = 9;
        while right < s.len() {
            let mut tmp = vec![];
            for i in left..=right {
                tmp.push(s[i]);
            }
            if set.contains(&tmp) {
                ans1.insert(tmp.clone());
            }
            set.insert(tmp);
            left += 1;
            right += 1;
        }
        let ans = ans1.into_iter().collect::<Vec<_>>();
        let mut result = vec![];
        for item in ans.into_iter() {
            result.push(item.iter().collect::<String>())
        }

        result
    }
}

File: src/187_Repeated_DNA_Sequences.py
# 187_Repeated_DNA_Sequences
# https://leetcode.cn/problems/repeated-dna-sequences/description/?envType=daily-question&envId=2023-11-05

from typing import List
class Solution:
    def findRepeatedDnaSequences(self, s: str) -> List[str]:
        if len(s) <= 10:
            return []
        l = 0
        r = 10
        st = set()
        ans = []
        while r <= len(s):
            if s[l:r] in st:
                ans.append(s[l:r])
            else:
                st.add(s[l:r])
            l += 1
            r += 1
        
        return list(set(ans))
        
s = Solution()
ans = s.findRepeatedDnaSequences("AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT")
print(ans)


from collections import defaultdict
from typing import List

class Solution:
    def findRepeatedDnaSequences(self, s: str) -> List[str]:
        # DNA sequence length
        L = 10
        
        # Mapping each nucleotide to a 2-bit binary representation
        bin_map = {
            'A': 0,  # 00
            'C': 1,  # 01
            'G': 2,  # 10
            'T': 3   # 11
        }
        
        n = len(s)
        
        # If the sequence is shorter than 10, return an empty list
        if n <= L:
            return []
        
        result = []
        rolling_hash = 0
        
        # Initialize the first L-1 nucleotides' rolling hash
        for ch in s[:L - 1]:
            rolling_hash = (rolling_hash << 2) | bin_map[ch]
        
        sequences_count = defaultdict(int)
        
        for i in range(n - L + 1):
            # Slide the window to consider the next nucleotide for hashing
            rolling_hash = ((rolling_hash << 2) | bin_map[s[i + L - 1]]) & ((1 << (L * 2)) - 1)
            
            # Increase the count of this hash
            sequences_count[rolling_hash] += 1
            
            # If the sequence appears twice, add it to the result
            if sequences_count[rolling_hash] == 2:
                result.append(s[i: i + L])
        
        return result
